Support GNUPGHOME environment variable

dennisdegreef opened the pull request:
I sometimes use the `GNUPGHOME` variable to point to a different set of rings.

This pull request should enable the usage of the environment variable and perform the previous behaviour if it is not set.
Let me know what you think.

dennisdegreef commented:
I'm also looking into writing a test for this usecase. Currently I have this:
```
func (s *LocalPGPTest) TestGnupgHomeOverride() {
	os.Setenv("GNUPGHOME", "/foo")
	_, err := NewLocalPGPService()
	s.True(err.Error() == "stat /foo/.gnupg/pubring.gpg: no such file or directory")
	os.Unsetenv("GNUPGHOME")
}
```

Which I find a bit ugly, but else you'd have to mock out the `os.Stat` call. Any thoughts on this?

ellotheth commented:
Great idea to use `$GNUPGHOME`. I'll be able to dig into this over the weekend and get you more feedback. Thanks!

ellotheth commented on lookup/localpgp.go:
```diff
@@ -29,7 +29,15 @@ type LocalPGPService struct {
+		home = os.Getenv("GNUPGHOME")
+	}
+
+	ringfile := path.Join(home, ".gnupg", "pubring.gpg")
```
If `$GNUPGHOME` is set, `ringfile` here would be `$GNUPGHOME/.gnupg/pubring.gpg`. Is that the standard usage? Reading the `gpg` man page, it looks like `$GNUPGHOME` is equivalent to `$HOME/.gnupg`, i.e. the top-level GnuPG directory:

> --homedir dir
>        Set the name of the home directory to dir. If this option is not used, the home directory defaults to `~/.gnupg`.  It is only recognized when given on the command line.  It also overrides any  home  directory  stated  through  the  environment  variable `GNUPGHOME`

dennisdegreef replied:
You are right, I could have tested this. I don't use $GNUPGHOME on the machine I coded this :)
I rebased my branch from master and changed the path. Should be correct now.

ellotheth commented:
For the test, it might make sense to extract the `ringfile` composition out to a method that returns the full path, and then test the method instead of the whole constructor. That way you don't have to worry about testing whether the file exists with `os.Stat`. (Which is really pushing off the inevitable, but at least then it's my problem, not yours!)

dennisdegreef commented:
Hi @ellotheth 

I just refactored the ringfile to a separate struct and added/changed tests for multiple use cases.
As for pushing off the os.Stat mocking, I think splitting for now is a good solution, since we want to primarily test the logic in case the environment variable is present, not if the os.Stat call is working.

https://github.com/dennisdegreef/pipethis/pull/1/files

Let me know what you think, when I merge this, it'll be added in this branch/PR.

ellotheth commented:
Hi @dennisdegreef,

Your branch looks stellar, but I'm not quite ready to start wrapping `os` methods to that extent. I had in mind something simpler, like https://github.com/ellotheth/pipethis/compare/e9995dc...alt-for-gnupghome. It skips dealing with `os.Stat` entirely, and focuses on building the file path. What do you think?

dennisdegreef commented:
It looks good! :+1: I was just trying out a different solution, while making myself more familiar with Golang.
Your solution is indeed simpler. I'll cherry-pick that commit in my branch. Thanks.

ellotheth commented:
Now that I've dirtied up `master`, do you mind rebasing your branch? I'll merge and tag everything afterwards.

dennisdegreef commented:
No problem, pushed it
